### DISCUSSION TITLE
FreeBSD Build

### DISCUSSION THREAD

**Discussion by spec1re** [StoneStepsInc/StoneStepsWebalizer#7] [Ideas]
Hi, thanks for such a great product! I'm using StoneStepsWebalizer for IIS to check the Exchange OWA logs, which gives a really nice overview, now I would like to do the same for our Squid logs on FreeBSD but there is just the "old" version available:

https://www.freshports.org/www/webalizer/

I can compile some packages my self, but webalizer with all his dependencies is a little over my head.

It would be very nice to have a build of StoneStepsWebalizer for FreeBSD:12:amd64 available. 

Thanks.

**Comment by StoneStepsInc** [comment]
I don't think I tried building on a FreeBSD VM. Will need to set one up and give it a go to see how it builds and whether Azure DevOps can work with it. Will post an update here, probably in a couple of weeks or so.

**Reply by spec1re** [reply]
That sound great! Take your time no need to hurry. At the moment I'm transferring the squid logs over to the windows machine and run the webalizer there, but this isn't optimal.

**Comment by StoneStepsInc** [comment]
FreeBSD turned out a bit more of a challenge than I expected. I only built it on FreeBSD 12, so if you are using anything earlier, it may not work because the build requires Clang v11 (C++ compiler), which I'm not sure is available on earlier versions of FreeBSD, but I didn't check. This is because of a bug in Clang versions prior to v11. I tried GCC that implemented required functionality in their earlier version, but it just crashes on FreeBSD, so I set it up with Clang.

You can run two scripts on FreeBSD 12 to build SSW v6.0.0 -  `freebsd-pkgs` installs all FreeBSD packages and `freebsd-build` grabs the source from the `v6-0` branch and builds it. You don't need to copy the source - the latter script will do this. Here are both:

https://raw.githubusercontent.com/StoneStepsInc/StoneStepsWebalizer/v6-0/devops/freebsd-pkgs
https://raw.githubusercontent.com/StoneStepsInc/StoneStepsWebalizer/v6-0/devops/freebsd-build

Copy both to any directory, `chmod a+x` both and run `freebsd-pkgs` first. Have a look at the source first - it doesn't ask any questions and just installs packages.

`freebsd-build` installs `wget`, if it's not installed (`pkg` will ask you to confirm), and grabs the source from the `v6-0`  branch. After it finishes, if there are no errors, there will be a directory `webalizer-6-0` created in the current directory and a package file `webalizer-6-0-0-7.tar.gz` inside it with all the usual package content.

You can uninstall `bash`, `gmake`, `googletest` after this. Other packages seem to combine development and runtime libraries, so they probably have to stay.

The source patch it applies changes the configuration directory to `/usr/local/etc` - seems to be the rule on FreeBSD. If you need `/etc`, you will need to change `ETCDIR` in `Makefile` and run `gmake` and `gmake package` again.

I didn't do much testing with this build - just ran it against a couple of logs to see if it works. Seems to produce what's expected, but I didn't go beyond that.

Lastly, state databases between different OS builds shouldn't be shared (i.e. `webalizer.db` and the renamed ones). Keep those generated on FreeBSD separate in case if you want to generate new reports from them.

**Reply by spec1re** [reply]
Wow, will check it out later!

**Comment by spec1re** [comment]
Excellent! 

```
Stone Steps Webalizer v6.0.0.7 (FreeBSD 12.2-STABLE)

Processed language file /usr/local/www/webalizer/lang/webalizer_lang.german
Fahre mit Konfigurationsdatei fort /usr/local/etc/webalizer.conf
Benutze Datei für den DNS-Cache /usr/local/www/webalizer/dns_cache.db
Benutze GeoIP-Datenbank /usr/local/www/webalizer/maxmind/GeoLite2-City.mmdb
Using ASN database /usr/local/www/webalizer/maxmind/GeoLite2-ASN.mmdb
DNS-Abfrage (20)
Using database /usr/local/www/webalizer/webalizer.db
Lese Datei mit historischen Daten... /usr/local/www/webalizer/webalizer.hist
Erzeuge Dateien in /usr/local/www/webalizer
Rechnername für Bericht ist 'SQUID'
Verwende Protokolldatei /var/squid/logs/access.log (Squid)
Finished log file /var/squid/logs/access.log
Speichere Daten des aktuellen Programmlaufes... [03/22/2021 08:36:36]
Speichere historische Daten...
Erzeuge Bericht für Maerz 2021 (HTML)
Erzeuge Zusammenfassung (HTML)
Processed 42133 Einträge (41759 ignoriert) in 0.18 Sekunden
DNS wait time is 0.00 Sekunden
Generated reports in 1.46 Sekunden
Maintenance time is 0.01 Sekunden
Total run time is 1.65 Sekunden
[root@squid]/root:
```
Compiling was easy and without any errors and no bugs so far.

If you like you can commit the build to the FreeBSD repo, so everyone can install it via pkg. 

Thanks again, much appreciated!

**Reply by StoneStepsInc** [reply]
Thanks for the details. Always good to know how it went.

I don't think it's quite ready for the repo. I would have to figure out how to build all Linux distribution without having to patch the source. The code in the trunk also has install/uninstall scripts and those would have to be adjusted to accommodate FreeBSD's different file hierarchy. Not having Docker on FreeBSD also makes things more complicated for building and testing.
